{% extends "base.html" %}
{% load staticfiles %}

{% block includes %}
<script type="text/javascript">
$(function() {
  $('#help_navbar_button').addClass('active');      
});
</script>
{% endblock %}

{% block content %}
<div class="container" style="margin-top:15px;">
  <div class="row">
    <div class="col-xs-2">
      <ul class="nav nav-pills nav-stacked">
        <li role="presentation" class="active"><a href="#background" aria-controls="background" role="tab" data-toggle="tab">Background</a></li>
        <li role="presentation"><a href="#contact" aria-controls="contact" role="tab" data-toggle="tab">Contact</a></li>
        <li role="presentation"><a href="#nomenclature" aria-controls="nomencalture" role="tab" data-toggle="tab">Nomenclature</a></li>
        <li role="presentation"><a href="#classification" aria-controls="classification" role="tab" data-toggle="tab">Classification</a></li>
        <li role="presentation"><a href="#faq" aria-controls="faq" role="tab" data-toggle="tab">FAQ</a></li>
        <li role="presentation"><a href="#cite" aria-controls="cite" role="tab" data-toggle="tab">How To Cite</a></li>
      </ul><br />
      <div class="center-block">
        <a class="github-button" href="https://github.com/edraizen/HistoneDB" data-style="mega" data-count-href="/edraizen/HistoneDB/watchers" data-count-api="/repos/edraizen/HistoneDB#subscribers_count" data-count-aria-label="# watchers on GitHub" aria-label="Watch edraizen/HistoneDB on GitHub" style="width:100%;">Watch</a><br />
        <a class="github-button" href="https://github.com/edraizen/HistoneDB/issues" data-icon="octicon-issue-opened" data-style="mega" data-count-api="/repos/edraizen/HistoneDB#open_issues_count" data-count-aria-label="# issues on GitHub" aria-label="Issue edraizen/HistoneDB on GitHub">Issue  </a>
      </div>
    </div>
    <div class="col-xs-10">
      <div class="tab-content">
        <div role="tabpanel" class="tab-pane active" id="background">
          <h2 id="background">Background</h2>
          Nucleosomes represent the basic unit of chromatin packing and wrap 145–147 bpof DNA in a 1.7 left-handed super helical turns around
a histone octamer composed of four types of core histones (H3,
H4, H2A, H2B — two copies of each). Nucleosomes are located at certain distances from each other along the DNA molecule and a linker histone H1 interacts with the linker DNA to form higher order chromatin structures. Nucleosomes are key players in epigenetic regulation of transcription,
replication and reprogramming; they may change their conformation, histone content, employ a set of different histone variants and become post-translationally modified upon certain conditions.
        </div>
        <div role="tabpanel" class="tab-pane" id="contact">
          <h2 id="contact">Contact</h2>
          <div class="container">
            <div class="row">
              <div class="col-xs-12">
                <!-- Place this tag where you want the button to render. -->
                <!-- Place this tag where you want the button to render. -->
                <a class="github-button" href="https://github.com/edraizen/HistoneDB" data-style="mega" data-count-href="/edraizen/HistoneDB/watchers" data-count-api="/repos/edraizen/HistoneDB#subscribers_count" data-count-aria-label="# watchers on GitHub" aria-label="Watch edraizen/HistoneDB on GitHub">Watch</a>
                <a class="github-button" href="https://github.com/edraizen/HistoneDB/issues" data-icon="octicon-issue-opened" data-style="mega" data-count-api="/repos/edraizen/HistoneDB#open_issues_count" data-count-aria-label="# issues on GitHub" aria-label="Issue edraizen/HistoneDB on GitHub">Issue</a>
              </div>
            </div>
            <div class="row">
{#              <div class="col-xs-4 center-block">#}
{#                <h3><a href="http://www.ncbi.nlm.nih.gov/CBBresearch/Panchenko/">Anna Panchenko</a></h3><br />#}
{#                Computational Biophysics Group<br />#}
{#                NIH, NLM, CBB<br />#}
{#                8600 Rockville Pike, Bldg 38A<br />#}
{#                Bethesda, MD, 20894<br />#}
{#                email here<br />#}
{#              </div>#}
              <div class="col-xs-4 center-block">
                <h3>Eli Draizen</h3><br />
                Computational Biophysics Group<br />
                NIH, NLM, CBB<br />
                8600 Rockville Pike, Bldg 38A<br />
                Bethesda, MD, 20894<br />
                eli.draizen _at_ nih.gov<br />
              </div>
              <div class="col-xs-4 center-block">
                <h3>Alexey K. Shaytan</h3><br />
                Computational Biophysics Group<br />
                NIH, NLM, CBB<br />
                8600 Rockville Pike, Bldg 38A<br />
                Bethesda, MD, 20894<br />
                alexey.shaytan _at_ nih.gov<br />
              </div>
            </div>
{#            <div class="row">#}
{#              <div class="col-xs-4 center-block">#}
{#                <h3><a href="http://www.ncbi.nlm.nih.gov/research/staff/landsman/">David Landsman</a></h3><br />#}
{#                Bioinformatics of Chromatin Structure Group<br />#}
{#                NIH, NLM, CBB<br />#}
{#                8600 Rockville Pike, Bldg 38A<br />#}
{#                Bethesda, MD, 20894<br />#}
{#                email here<br />#}
{#              </div>#}
{#              <div class="col-xs-4 center-block">#}
{#                <h3><a href="http://www.ncbi.nlm.nih.gov/research/staff/marino/">Leonardo Mariño-Ramírez</a></h3><br />#}
{#                Bioinformatics of Chromatin Structure Group<br />#}
{#                NIH, NLM, CBB<br />#}
{#                8600 Rockville Pike, Bldg 38A<br />#}
{#                Bethesda, MD, 20894<br />#}
{#                email here<br />#}
{#              </div>#}
{#          </div>#}
        </div>
      </div>
      <div role="tabpanel" class="tab-pane" id="nomenclature">
        <h2 id="nomencalture">Nomenclature</h2>
        Histone variants need to be .. <br /><br />
        [desc] [prefix]<font color="red">TYPE</font>.<font color="blue">VARIANT</font>.<font color="orange">GENE</font>.<font color="pink">sSPLICE</font> <br />
        <ol>
          <li><font color="red">TYPE</font> = Core histone name (H2A, H2B, H3, H4, H1)</li>
          <li><font color="blue">VARIANT</font>:
            <ul>
              <li>Letter: "Structurally distinct monophyletic clade of a histone family", exception is H2A.X</li>
              <li>Number: Species-specific variant, not defined by phylogeny but clear orthology</li>
            </ul>
          </li>
          <li><font color="orange">GENE</font> = Gene number, or more specially phylogenetic branch point order (Optional if unknown)</li>
          <li><font color="pink">SPLICE</font> = Splice isoform index, or paralogous sequences, preceded by ‘s’ character (Optional if unknown)</li>
        </ol>
        prefix = Structurally distinct clade of core histone (<b>macro</b>H2A, <b>sub</b>H2B, <b>cen</b>H3)<br />
        desc = Specify features about variant (location, species)<br />
        Note:<br />
        <ul>
          <li>Capitalization</li>
          <li>Periods (Phylogenetic branch points)</li>
          <li>Spaces (Separate description from variant description)</li>
        </ul>
        <div class="container text-center">Example: <font color="red">H2A</font>.<font color="blue">Z</font>.<font color="orange">1</font>.<font color="pink">s1</font></div>
      </div>
      <div role="tabpanel" class="tab-pane" id="classification">
       Curated alignments of histone variants were used to train Hidden Markov Models, utilizing HMMER 3.1b2,
          to create one HMM for each variant. These models were used as a part of automatic extraction and annotation
          pipeline. Namely, all sequences from the nr protein database have been classified by the HMM models
          (variants and canonical) and were added into the HistoneDB 2.0 as “automatically extracted sequences”.
          We assigned a model with a maximum HMMER score to a given sequence. The score was required to exceed
          a certain threshold identified as follows. For any given variant model, we searched the curated sequence
          set and calculated HMMER scores for all curated sequences. We then varied the HMMER score threshold
          in order to distinguish variant sequences from all other sequences with 90% specificity. Specificity
          of retrieval was estimated based on the number of true positives (TPs; sequences correctly predicted by
          their native model) and false positives (FPs; incorrectly predicted sequences) found above each HMMER
          score cutoff. The specificity was calculated as TN/(FP+TN). Then a desired score cutoff value was
          obtained from the interpolated inverse curve of the score cutoffs plotted versus specificity. The
          efficacy of classification based on HMM models depends significantly on whether the respective
          histone variant model is sufficiently divergent and forms a separate clade on the phylogenetic
          tree. For certain variants,
          which emerged repeatedly in the course of evolution, the classification becomes difficult unless a
          characteristic signature is known. In the case of H2A.X, HMM search algorithm was supplemented by the
          pattern matching, since this variant is characterized by the presence of ‘SQ(E/D)Φ’ motif as described
          above. Any sequence that was classified as canonical H2A by HMMER-based algorithm was reclassified as
          H2A.X if it had this motif at its C-terminus. The described above classification algorithm can be also
          applied to classify any sequence of interest.
        </div>
      <div role="tabpanel" class="tab-pane" id="faq">
          <div class="container">
            <div class="row">
              <h2>FAQ</h2>
            </div>
            <div class="row">
              <h3>Who do I report bugs to?</h3>
            </div>
            <div class="row">
              Please add an issue to github with the bug tag! <a class="github-button" href="https://github.com/edraizen/HistoneDB/issues" data-icon="octicon-issue-opened" data-style="mega" data-count-api="/repos/edraizen/HistoneDB#open_issues_count" data-count-aria-label="# issues on GitHub" aria-label="Issue edraizen/HistoneDB on GitHub">Issue</a>
            </div>
            <div class="row">
              <h3>Can you add XXXX feature?</h3>
            </div>
            <div class="row">
              Please add an issue to gihub with the enhancement tag! <a class="github-button" href="https://github.com/edraizen/HistoneDB/issues" data-icon="octicon-issue-opened" data-style="mega" data-count-api="/repos/edraizen/HistoneDB#open_issues_count" data-count-aria-label="# issues on GitHub" aria-label="Issue edraizen/HistoneDB on GitHub">Issue</a>
            </div>
            <div class="row">
              <h3>How can I download all sequences?</h3>
            </div>
            <div class="row">
              We are still working on this.
            </div>
          </div>
      </div>

      <div role="tabpanel" class="tab-pane" id="cite">
        Eli J. Draizen, Alexey K. Shaytan, Leonardo Marino-Ramirez, Paul B. Talbert, David Landsman, Anna R. Panchenko. "The HistoneDB 2.0: a phylogeny-based resource for all histone proteins and their variants." <i>Database</i>. In Preparation.
      </div>
    </div>
  </div>
</div>
<!-- Place this tag right after the last button or just before your close body tag. -->
<script async defer id="github-bjs" src="https://buttons.github.io/buttons.js"></script>
{% endblock %}
